### ISSUE TITLE
Improve extender API

### ISSUE THREAD

**Issue by pablo-abc** [pablo-abc/felte#30]
While the current extender API works, it feels quite hacky. A major thing that should be available is to modify or extend the Felte's configuration which currently relies on modifying the configuration object taking advantage of its mutability.

**Comment by benbender** [comment]
Hey,

as said i #43 I had a quick look at the extender-/validator-API to use the CVAPI for validation and felte for formstate and rendering the error-messages. What I first stumbled over was the fact, that the form isn't available to neither the validate-API nor extender-API (as it was null at my tests).

I'm a bit unsure if the nullified currentform was intentional in the design or a consequence (seemed to be the later as all the other implementations, I had a sneak at, did some sort of null-checking in the extender-api at the beginning). 

If you don't want to pass the form as a second argument to validate() for separation-concerns it would be nice to have the form-object at least available to the setup-phase of the extend-API to f.e. add extra eventhandlers for validation-events programmatically.

Reasoning: Atm my use-case, detailed in #43, doesn't seem to be possible by using the felte-apis in the intended way without  doing much hackery. But I also have to say that the use-case is clearly an edge-case which might not be worth to lift the barriers introduced by the correct, nice & clear separation of concerns in the APIs...

**Comment by pablo-abc** [comment]
It is not available to `validate` for the reasons mentioned previously. While, in the case of the extender API, it is not available during the _first_ call because it happens during component initialisation, not _mount_. The reason for this is that you might want to set up features during component initialisation (such as context as it's done by `reporter-svelte` and `reporter-solid`). This is why you see null checks in some extenders with `if (!currentForm.form) return;`.

Going back to #43, I understand your use case. There's a few things to consider here:

1. The browser won't even set a forms validity if the form has the `novalidate` attribute. So `checkValidity` will always return valid regardless of any attribute on the inputs.
2. If we remove the `novalidate` attribute then you remove control from Felte completely regarding validation. This will definitely cause conflicts and headaches.

But I do get we want mainly two things here:

1. Minimal repetition of validation (just by using native HTML attributes such as `required`).
2. Ability to still mix and match validators and reporters.

The main constraint I'd like to keep is to keep our stores (`data`, `errors`, `touched`) as the main _source of truth_ for all things. As an example I can show you the `reporter-cvapi` which _does_ use [`setCustomValidity`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLObjectElement/setCustomValidity). But the main source of truth for errors is still the `errors` store.

The flow of data right now looks more or less like this: `user input` -> `data store update` -> `touched store update` -> `validation` -> `error store update` -> `reporter calls`.

An API I think would solve your use case while still adhering to these constraints probably would look like:

```html
<script>
  import { validator } from '@felte/validator-cvapi';

  const { form } = createForm({
    // ...
    extend: [validator({ required: 'This field is required', \* ... *\  })],
    // ...
  });
</script>

<form>
  <input type="email" required />
  <input type="password" minlength={8} required />
  <button type="submit">Submit</button>
</form>
```

On which the `validator` would do most of the heavy lifting. But as far as I understand, as long as the form has `novalidate`, we would need the developer to handle messages and i18n, since we would not have access to the built-in browser ones.

**Comment by benbender** [comment]
Ah, I wasn't aware of the fact that the validation isn't running at all if you set `novalidate=true` on a form (even if it makes perfectly sense...). Nevertheless,  think there might be a better way to achive what I wanted.

There is an [invalid-event](https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement/invalid_event) which can be hooked on to to cancel the "bubble" and read the error-message/validity-state.

So it should be possible to do the following (if I'm not missing out a detail again):

1. set `novalidate=true` on init
2. hook on invalid to read the `validity`-state and `validationMessage` of an input-element and cancel the event. Use `validity`-state and `validationMessage` to update the errors-store of felte.
3. on the `validate()`-call of felte, set `novalidate=false` on the form, call `checkValidity()` on the form (or all inputs, not sure here) and set `novalidate=true` afterwards.

This should meet all requirements of my use-case and the principles of felte...

What do you think about this approach?

**Comment by benbender** [comment]
PS: to get back to the main-topic of the thread - the addition of a validator-function via extender-api should be handled by felte imho. Checking and combining them seems to be a bit repititive and cumbersome.

**Comment by pablo-abc** [comment]
That's actually quite ingenuous and much less hacky than I was expecting it to need to be. So as soon as `novalidate` is set to `false` the validation errors from the browser are available?

And yeah, as far as I understand the `invalid` event happens whenever you call `checkValidity()` on a field, which I thought you'd need to call before checking errors on the field... but it seems the browser sets the errors eagerly. I *really* like this approach, actually.

A couple of additions I'd think would be necessary are:

- For the sake of flexibility maybe we can let developers override messages by letting them pass an object with the same keys as [ValidityState](https://developer.mozilla.org/en-US/docs/Web/API/ValidityState)? Since design or business requirements might require specific messaging. Probably we'd need to pass information about the field in order for them to appropriately set the message (e.g. input `type` and current value).
- Instead of `cverrors[control.name] = ...`, let's use the `getPath` util from `@felte/common` like `cverrors[getPath(control)] = ...` since it handles possible nested fields.

The way I'm currently making validators that accept configuration is with higher order functions like `const extender => (config) => (currentForm) => { \*...*\};`

You up for making a PR with this validator? You can copy the files of another validator in the repo and change details accordingly.

**Comment by pablo-abc** [comment]
@benbender you're totally right. Currently modifying the configuration depends on the fact that you can mutate the `config` object since it's passed by reference. `addValidator` seems like a good first step towards not relying on that. I was also considering if there would ever need a reason for a dev to _override_ the configuration (for things like modifying `onSubmit` behaviour`?) or if this would always be purely additive.

**Comment by benbender** [comment]
> That's actually quite ingenuous and much less hacky than I was expecting it to need to be. So as soon as `novalidate` is set to `false` the validation errors from the browser are available?

yes

> And yeah, as far as I understand the `invalid` event happens whenever you call `checkValidity()` on a field, which I thought you'd need to call before checking errors on the field... but it seems the browser sets the errors eagerly. I _really_ like this approach, actually.

me too :) And yes, checkValidity() gets never called because the form is in `novalidate=true` on the submit-event. Thats why the solution works that simple.

> A couple of additions I'd think would be necessary are:

Sure!

> * For the sake of flexibility maybe we can let developers override messages by letting them pass an object with the same keys as [ValidityState](https://developer.mozilla.org/en-US/docs/Web/API/ValidityState)? Since design or business requirements might require specific messaging. Probably we'd need to pass information about the field in order for them to appropriately set the message (e.g. input `type` and current value).

I'm a bit unsure about this as the config might get quite complicated. Maybe a better solution would be to offer to configure a callback per field which calls `setCustomValidity()` for the field on error?

> * Instead of `cverrors[control.name] = ...`, let's use the `getPath` util from `@felte/common` like `cverrors[getPath(control)] = ...` since it handles possible nested fields.

Ah, sure. Didn't know about `getPath()`.

> The way I'm currently making validators that accept configuration is with higher order functions like `const extender => (config) => (currentForm) => { \*...*\};`

👍 

> You up for making a PR with this validator? You can copy the files of another validator in the repo and change details accordingly. (It doesn't need to be ready to merge, you may set the PR as a `draft`).
Sure, but it may take some days as I have some paid work to do and have been fiddeling around with this for too long this morning :)

And we'll have to see if tests are easily possible as I'm a bit unsure about the validation-api-support in jsdom. And I would like to leave up the docs up to you as you know your style and the structures of your docs.

> you're totally right. Currently modifying the configuration depends on the fact that you can mutate the config object since it's passed by reference. addValidator seems like a good first step towards not relying on that. I was also considering if there would ever need a reason for a dev to override the configuration (for things like modifying onSubmit behaviour`?) or if this would always be purely additive.

My take here: don't try to overengineer. I think the api is basically fine & easy to understand. Solve problems which are real. Premature optimization is the root of all evil ;)

**Comment by benbender** [comment]
And, last but not least, two questions about the implementation (see comments):
```ts
const validatorCvapiExtender = (
    currentForm: CurrentForm<Record<string, string>>
): ExtenderHandler<Record<string, string>> => {
  // Is the "currentForm.config.cvapivalidation"-check neccessary?
  // The extender gets called at least twice and I wanted to ensure to not
  // add it multiple times. Is this neccessary?
  // If it's needed, is there a better place to store this peace of state other
  // than the config-object?
  if (currentForm.form && currentForm.config.cvapivalidation !== true) {
    const validate = currentForm.config.validate;
    const validateFn: ValidationFunction<Record<string, string>> = () => {
      const cverrors: Errors<Record<string, string>> = {};

      // can this check live outside of the validate()-function?
      if (currentForm.form && currentForm.controls) {
        currentForm.form.novalidate = false;

        currentForm.controls.forEach((control) => {
          cverrors[getPath(control)] = control.validationMessage;
        });

        currentForm.form.novalidate = true;
      }

      return cverrors;
    };

    if (validate && Array.isArray(validate)) {
      currentForm.config.validate = [...validate, validateFn];
    } else if (validate) {
      currentForm.config.validate = [validate, validateFn];
    } else {
      currentForm.config.validate = [validateFn];
    }

    // see above.
    currentForm.config.cvapivalidation = true;
  }

  return {};
};
```

**Comment by benbender** [comment]
First preview of `validator-cvapi` available here: https://github.com/benbender/felte/tree/feat/validator-cvapi

**Comment by pablo-abc** [comment]
The reason I'd prefer a single object with each validity state as property is that it's a more common case to have common error messages for all of your validations than a custom error per field. So more most use-cases it'd require little duplication (which I'd also think is better DX?). Similar to how it works in Yup, as an example. So probably accepting either a string or a function that'll receive information about the field that failed would work for better DX in general (unless there's something I'm not seeing?).

---

The thing regarding your `cvapivalidation` brings another issue with the current extender API. Maybe having defined lifecycles for an extender would make more sense. As of now there's three situations on which extenders are called:

1. On component initialisation (when `form == null`).
2. On mount (when the `form` action in `use:form` gets called).
3. Whenever the form changes (e.g. fields have been added or removed from the DOM).

As of now, no package had to differentiate from `mount` and `update`, so I only needed to check for the existence of `form` to different from initialisation and the rest. So yes, for now that seems to be the only way to differentiate from an update and the first `mount` call. But it does stay as a note to myself to add a proper way to handle lifecycles.
